### PR TITLE
clean up of spack clean

### LINKED ISSUE #8608: spack clean --python-cache deletes .pyc files in installed packages
I just noticed that `spack clean --python-cache` deletes `.pyc` files in Spack-installed packages. If you install Python or Python-packages with Spack and then run `spack --debug clean --python-cache`, you can reproduce this.

We should probably ignore the `opt/` directory when traversing the filesystem. I don't think it will hurt anything, but it does slow down the Python installation temporarily and cause `spack clean` to take longer than it should.

Also, the `help` message for `--all` needs to be updated with the `-p` flag.

P.S. Can we finally delete https://github.com/spack/spack/blob/develop/bin/spack#L59,L81? These files haven't existed for years, and now we have a simple solution to clean them if users run into trouble.

@healther

### PR DESCRIPTION
**Pull request by healther**
closes #8608 

* update help of `clean --all` to include `-p`
* remove old orphaned .pyc removal

@adamjstewart will the first line work this way?

### CODE REVIEW

**Comment by adamjstewart** [comment]
(2) is already the current behavior, it just updates the help message to reflect this. I think it is the right behavior. It can't possibly hurt, but it can possibly help. If someone encounters a problem, `spack clean -a` should be sufficient to solve that problem.

I agree on only cleaning up Spack files, not installation files. I don't mind having (1) in the same PR, but it's up to you.

**Comment by tgamblin** [comment]
Ok I think we can leave `-p` in `-a` for now. I think we’ll need to revisit `spack clean` if we rework permisssions on spack, e.g. so that users don’t have to have write access on the installation directory to use it (something I want to do).

I’m ok with what’s in this PR but it should also make clean not touch things in the `install_tree`, or things in Spack’s prefix that Spack doesn’t own (things not in `$prefix/*/spack`).

**Comment by healther** [comment]
I wasn't aware that `setup.py install` already directly produces `.pyc` files -> removing `SPACK_ROOT/opt` from cleanup

**Comment by healther** [comment]
@tgamblin ping

**Comment by adamjstewart** [comment]
I don't think the current implementation handles:

> things in Spack’s prefix that Spack doesn’t own (things not in `$prefix/*/spack`).

Spack has a `spack clone` command that creates a copy of Spack in a directory like `/usr/local`. If there are other things installed in `/usr/local`, it isn't sufficient to ignore `opt`, there may be other non-Spack related files in there.

**Comment by healther** [comment]
What are these
> things in Spack’s prefix that Spack doesn’t own (things not in $prefix/*/spack).

?

**Comment by adamjstewart** [comment]
@healther see my above comment

File: lib/spack/spack/cmd/clean.py [hunk 1]
```diff
@@ -98,6 +98,8 @@ def clean(parser, args):
     if args.python_cache:
         tty.msg('Removing python cache files')
         for root, dirs, files in os.walk(spack_root):
```
Comment (adamjstewart) [inline lib/spack/spack/cmd/clean.py]: I think if you replace this line with:
```python
from spack.paths import lib_path, var_path

for spack_root in [lib_path, var_path]:
    for root, dirs, files in os.walk(spack_root):
```
you should be good to go. Those seem to be the only directories containing anything that gets imported.

**Comment by healther** [comment]
sorry for taking so long. Work is kind of busy right now...

I wasn't even aware of `spack clone`. Do you know what the intended use case is? Is this "new spack" in any way associated with the "current spack"? 

btw. I think the help of `spack clone` contains a typo (see #8658)

**Comment by tgamblin** [comment]
> I wasn't even aware of spack clone. Do you know what the intended use case is? Is this "new spack" in any way associated with the "current spack"?

No. It's a way to bootstrap a new Spack installation in a system prefix.

**Comment by healther** [comment]
Doesn't 
```
    for spack_root in [lib_path, var_path]
```
do exactly that?

File: lib/spack/spack/cmd/clean.py [hunk 2]
```diff
@@ -97,14 +98,15 @@ def clean(parser, args):
...
-                if f.endswith('.pyc') or f.endswith('.pyo'):
-                    fname = os.path.join(root, f)
-                    tty.debug('Removing {0}'.format(fname))
-                    os.remove(fname)
-            for d in dirs:
-                if d == '__pycache__':
-                    dname = os.path.join(root, d)
-                    tty.debug('Removing {0}'.format(dname))
-                    shutil.rmtree(dname)
+        for folder in [lib_path, var_path]:
```
Comment (adamjstewart) [inline lib/spack/spack/cmd/clean.py]: I actually like `directory` better than `folder` if you're going to change the name. The docs for `os.walk` call it `top` if you want to be inline with that.

**Comment by tgamblin** [comment]
> Doesn't 
> ```
>     for spack_root in [lib_path, var_path]
> ```
> do exactly that?

Yep, duh!  Thanks!

**Pull request merged**